src/connector/src/source/google_pubsub/source/reader.rs

```diff
@@ -0,0 +1,166 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use core::result::Result::Ok as StdOk;
+
+use anyhow::{anyhow, ensure, Context, Ok, Result};
+use async_trait::async_trait;
+use chrono::{NaiveDateTime, TimeZone, Utc};
+use futures_async_stream::try_stream;
+use google_cloud_pubsub::client::Client;
+use google_cloud_pubsub::subscription::{SeekTo, Subscription};
+use risingwave_common::{bail, try_match_expand};
+use tonic::Code;
+
+use super::TaggedReceivedMessage;
+use crate::source::google_pubsub::PubsubProperties;
+use crate::source::{
+    BoxSourceStream, Column, ConnectorState, SourceMessage, SplitId, SplitImpl, SplitMetaData,
+    SplitReader,
+};
+
+const PUBSUB_MAX_FETCH_MESSAGES: usize = 1024;
+
+pub struct PubsubSplitReader {
+    subscription: Subscription,
+    split_id: SplitId,
+    stop_offset: Option<NaiveDateTime>,
+}
+
+impl PubsubSplitReader {
+    #[try_stream(boxed, ok = Vec<SourceMessage>, error = anyhow::Error)]
+    pub async fn into_stream(self) {
+        loop {
+            let pull_result = self
+                .subscription
+                .pull(PUBSUB_MAX_FETCH_MESSAGES as i32, None, None)
+                .await;
+
+            let StdOk(raw_chunk) = pull_result else {
```

Comment:
@tabVersion does everything look okay here?

Reply:
I don't think it is a good practice. Maybe we can use match here

```rust
match pull_result {
    Ok(chunk) => chunk,
    Err(e) => { ... do something ... }
}
```

Besides, if we get NotFound and PermissionDenied, we should throw the error rather than keep looping.

Reply:
We are returning an error on `NotFound` and `Permission` denied I think. I also think `match` would be better practice here, but is there a way to avoid nested `match` for the equivalent?

Reply:
Oh will we get other errors that don't affect the consuming process?
I thought any error could terminate the process.

Reply:
From what I saw other errors from the `pull` RPC are temporary, as in errors due to too many pending pulls or due to temporary failure of the service. Should we error out the reader on them, keep going or perhaps sleep for a bit before resuming?

Reply:
Then I think nested match is acceptable. Just ignore other errors

Reply:
Done and done